src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const STATES = [];
+  let copy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        copy = {};
+        break;
+      case 'addProperties':
+        copy = { ...copy, ...action.extraData };
+        break;
+      case 'removeProperties':
+        copy = { ...copy };
+
+        for (const key of action.keysToRemove) {
+          delete copy[key];
+        }
+        break;
+      default:
+        break;
+    }
+
+    STATES.push(copy);
```

Comment:
[CODE STYLE] - It is better to push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after current action is processed. This way, you ensure that the state is only added to the 'STATES' array after an action has been fully applied.